### PR TITLE
Added FAQ / helpURL at widgets

### PR DESCRIPTION
**Pull request by Baertram**
[Widget versions were not yet increased! Please do  so if neccessary]
-Added attribute helpUrl to most widgets
-Added functions to LibAddonMenu-2.0.lua to create a FAQ texture (?) control at the widget controls (next right of their label text's width, or if no label is given right of the widget control)
-Hovering the texture will show "Open website: " .. url
-Clicking the texture will open the URL open dialog of ESO

### CODE REVIEW

File: LibAddonMenu-2.0/LibAddonMenu-2.0/LibAddonMenu-2.0.lua [hunk 1]
```diff
@@ -78,10 +78,73 @@ local function GetStringFromValue(value)
...
+    local onMouseExitAlpha = 0.4
+
+    local helpUrl = controlData and controlData.helpUrl and GetStringFromValue(controlData.helpUrl)
+    if not helpUrl or helpUrl == "" then return end
+    local faqControl = wm:CreateControl(nil, parent, CT_TEXTURE)
+    faqControl:SetHidden(false)
+    faqControl:ClearAnchors()
+    faqControl:SetDrawLayer(DL_OVERLAY)
+    faqControl:SetTexture("\\ESOUI\\ART\\miscellaneous\\help_icon.dds") --small (i): \\ESOUI\\ART\\help\\help_tabicon_tutorial_up.dds
+    if controlData.isHalfWidth then
```
Comment (sirinsidiator) [inline LibAddonMenu-2.0/LibAddonMenu-2.0/LibAddonMenu-2.0.lua:103]: Is this really needed? Can't they just have the same size in both cases?

File: LibAddonMenu-2.0/LibAddonMenu-2.0/LibAddonMenu-2.0.lua [hunk 2]
```diff
@@ -78,10 +78,73 @@ local function GetStringFromValue(value)
...
+        faqControl:SetDimensions(21, 21)
+    else
+        faqControl:SetDimensions(23, 23)
+    end
+
+    faqControl:SetColor(FAQ_ICON_COLOR:UnpackRGBA())
+    faqControl:SetAlpha(onMouseExitAlpha)
+    faqControl:SetAnchor(anchorPoint, anchorRelativeControl, anchorRelativePoint, anchorX, anchorY-1.5)
```
Comment (sirinsidiator) [inline LibAddonMenu-2.0/LibAddonMenu-2.0/LibAddonMenu-2.0.lua:111]: Do decimal offsets exist? I only ever saw and used integer values until now.

File: LibAddonMenu-2.0/LibAddonMenu-2.0/LibAddonMenu-2.0.lua [hunk 3]
```diff
@@ -78,10 +78,73 @@ local function GetStringFromValue(value)
...
+
+    faqControl.isMouseOver = false
+    faqControl.data = faqControl.data or {}
+    faqControl.data.helpUrl = helpUrl
+    faqControl.data.tooltipText = util.L.WEBSITE .. ": " .. helpUrl
```
Comment (sirinsidiator) [inline LibAddonMenu-2.0/LibAddonMenu-2.0/LibAddonMenu-2.0.lua:116]: Maybe use string.format with a template defined in a constant variable. e.g. HELP_URL_TOOLTIP:format(...)

File: LibAddonMenu-2.0/LibAddonMenu-2.0/LibAddonMenu-2.0.lua [hunk 4]
```diff
@@ -78,10 +78,73 @@ local function GetStringFromValue(value)
...
+        end
+    end)
+    faqControl:SetHandler("OnMouseEnter", function(self)
+        self.isMouseOver = true
+        ZO_Options_OnMouseEnter(self)
+        self:SetColor(FAQ_ICON_MOUSE_OVER_COLOR:UnpackRGBA()) --light blue
+        self:SetAlpha(1)
+    end, "LAM2_FAQTexture_OnMouseEnter")
+    faqControl:SetHandler("OnMouseExit", function(self)
+        onMouseExitFAQ(self)
```
Comment (sirinsidiator) [inline LibAddonMenu-2.0/LibAddonMenu-2.0/LibAddonMenu-2.0.lua:142]: Just feed onMouseExitFAQ into SetHandler directly instead of wrapping it.

File: LibAddonMenu-2.0/LibAddonMenu-2.0/LibAddonMenu-2.0.lua [hunk 5]
```diff
@@ -121,10 +185,23 @@ local function CreateLabelAndContainerControl(parent, controlData, controlName)
         label:SetAnchor(TOPRIGHT, container, TOPLEFT, 5, 0)
     end
 
+    local anchorFAQX = 0
+    local anchorFAQPos = LEFT
+    local anchorFAQRelativePos = LEFT
+    if label.GetTextWidth then
```
Comment (sirinsidiator) [inline LibAddonMenu-2.0/LibAddonMenu-2.0/LibAddonMenu-2.0.lua:191]: Not sure what the purpose of this check is? Label should always be a label control, so it either has this method or not. Or is it a newly added function on the pts?

File: LibAddonMenu-2.0/LibAddonMenu-2.0/LibAddonMenu-2.0.lua [hunk 6]
```diff
@@ -121,10 +185,23 @@ local function CreateLabelAndContainerControl(parent, controlData, controlName)
...
+        anchorFAQPos = LEFT
+        anchorFAQRelativePos = RIGHT
+    end
+    CreateFAQTexture(control, controlData, controlName, anchorFAQPos, label, anchorFAQRelativePos, anchorFAQX, 0)
+
     control.data.tooltipText = GetStringFromValue(control.data.tooltip)
     control:SetMouseEnabled(true)
-    control:SetHandler("OnMouseEnter", ZO_Options_OnMouseEnter)
-    control:SetHandler("OnMouseExit", ZO_Options_OnMouseExit)
+    control:SetHandler("OnMouseEnter", ZO_Options_OnMouseEnter, "LAM2_Control_OnMouseEnter")
```
Comment (sirinsidiator) [inline LibAddonMenu-2.0/LibAddonMenu-2.0/LibAddonMenu-2.0.lua:202]: Is this change needed? If possible I'd like to avoid changing the current behaviour in unexpected ways. If anyone tries to hook the OnMouseEnter handler and expects to be able to prevent the LAM handler from running, it would break due to this.

File: LibAddonMenu-2.0/LibAddonMenu-2.0/controls/checkbox.lua [hunk 7]
```diff
@@ -102,8 +103,9 @@ end
 --controlName is optional
 function LAMCreateControl.checkbox(parent, checkboxData, controlName)
     local control = LAM.util.CreateLabelAndContainerControl(parent, checkboxData, controlName)
-    control:SetHandler("OnMouseEnter", OnMouseEnter)
-    control:SetHandler("OnMouseExit", OnMouseExit)
+    control:SetHandler("OnMouseEnter", OnMouseEnter, "LAM2_Control_OnMouseEnter")
```
Comment (sirinsidiator) [inline LibAddonMenu-2.0/LibAddonMenu-2.0/controls/checkbox.lua:106]: same consideration as above

File: LibAddonMenu-2.0/LibAddonMenu-2.0/controls/dropdown.lua [hunk 8]
```diff
@@ -382,8 +383,9 @@ function LAMCreateControl.dropdown(parent, dropdownData, controlName)
     local combobox = control.combobox
     combobox:SetAnchor(TOPLEFT)
     combobox:SetDimensions(control.container:GetDimensions())
-    combobox:SetHandler("OnMouseEnter", function() ZO_Options_OnMouseEnter(control) end)
-    combobox:SetHandler("OnMouseExit", function() ZO_Options_OnMouseExit(control) end)
+    combobox:SetHandler("OnMouseEnter", function() ZO_Options_OnMouseEnter(control) end, "LAM2_Control_OnMouseEnter")
```
Comment (sirinsidiator) [inline LibAddonMenu-2.0/LibAddonMenu-2.0/controls/dropdown.lua:386]: see above

File: LibAddonMenu-2.0/LibAddonMenu-2.0/controls/submenu.lua [hunk 9]
```diff
@@ -105,78 +106,80 @@ function LAMCreateControl.submenu(parent, submenuData, controlName)
...
-    animation:SetHandler("OnStop", function(self, completedPlaying)
-        scroll:SetResizeToFitDescendents(control.open)
-        if control.open then
-            control.arrow:SetTexture("EsoUI\\Art\\Miscellaneous\\list_sortup.dds")
-            scroll:SetResizeToFitPadding(5, 20)
-        else
-            control.arrow:SetTexture("EsoUI\\Art\\Miscellaneous\\list_sortdown.dds")
-            scroll:SetResizeToFitPadding(5, 0)
-            scroll:SetHeight(0)
+        control.scroll = wm:CreateControl(nil, control, CT_SCROLL)
```
Comment (sirinsidiator) [inline LibAddonMenu-2.0/LibAddonMenu-2.0/controls/submenu.lua:122]: why did the indentation change here?

File: LibAddonMenu-2.0/LibAddonMenu-2.0/LibAddonMenu-2.0.lua (hunk 1 above)
Comment (Baertram) [reply LibAddonMenu-2.0/LibAddonMenu-2.0/LibAddonMenu-2.0.lua:103]: I had changed it in size by 2 pixels to make it look and anchor properly for half width controls in submenus. Under some circumstances it was missplaced if the size was 23 pixels as well.
Couldn't rebuild the problem with my current tests thoug.
So I guess 23pixels would be okay for both.

File: LibAddonMenu-2.0/LibAddonMenu-2.0/LibAddonMenu-2.0.lua (hunk 3 above)
Comment (Baertram) [reply LibAddonMenu-2.0/LibAddonMenu-2.0/LibAddonMenu-2.0.lua:116]: I'm not sure if this would be an overkill for a simple string concatenation with 2 variables...

File: LibAddonMenu-2.0/LibAddonMenu-2.0/LibAddonMenu-2.0.lua (hunk 5 above)
Comment (Baertram) [reply LibAddonMenu-2.0/LibAddonMenu-2.0/LibAddonMenu-2.0.lua:191]: During my tests there was one circumstance where the GetTextWidth function did not exist on the "label" variable. Maybe I managed to pass some other control than a real label to it or it was a submenu's label (which got the function as well). Wil ldo some more tests.

File: LibAddonMenu-2.0/LibAddonMenu-2.0/LibAddonMenu-2.0.lua (hunk 6 above)
Comment (Baertram) [reply LibAddonMenu-2.0/LibAddonMenu-2.0/LibAddonMenu-2.0.lua:202]: Not sure why this should break? It's ZOs code and if this fails to load properly because there are PreHooks or PostHooks they should secure it or the devs need to adopt their code either way, I'd say.
Or I don't understand what the problem would be.

I had added this as I was using the OnMouseEnter -> Show/hide the faq texture. As you already wished the textrues to always show I can remove it again.

File: LibAddonMenu-2.0/LibAddonMenu-2.0/controls/submenu.lua (hunk 9 above)
Comment (Baertram) [reply LibAddonMenu-2.0/LibAddonMenu-2.0/controls/submenu.lua:122]: Dunno

File: LibAddonMenu-2.0/LibAddonMenu-2.0/controls/editbox.lua [hunk 10]
```diff
@@ -21,6 +22,14 @@ if not LAM:RegisterWidget("editbox", widgetVersion) then return end
 
 local wm = WINDOW_MANAGER
 
+local function GetValidTextType(textType)
```
Comment (sirinsidiator) [inline LibAddonMenu-2.0/LibAddonMenu-2.0/controls/editbox.lua:25]: Don't think this belongs into this pull request?

File: LibAddonMenu-2.0/LibAddonMenu-2.0/LibAddonMenu-2.0.lua [hunk 11]
```diff
@@ -1,1317 +1,1387 @@
...
+    faqControl:SetDimensions(23, 23)
+
+    faqControl:SetColor(FAQ_ICON_COLOR:UnpackRGBA())
+    faqControl:SetAlpha(onMouseExitAlpha)
+    faqControl:SetAnchor(anchorPoint, anchorRelativeControl, anchorRelativePoint, anchorX, anchorY-1)
+
+    faqControl.isMouseOver = false
+    faqControl.data = faqControl.data or {}
+    faqControl.data.helpUrl = helpUrl
+    --faqControl.data.tooltipText = HELP_URL_TOOLTIP:format(helpUrl)
```
Comment (sirinsidiator) [inline LibAddonMenu-2.0/LibAddonMenu-2.0/LibAddonMenu-2.0.lua:113]: Not sure if this is supposed to be commented out?
I also don't like how HELP_URL_TOOLTIP spreads this out into 3 places. Maybe just write `("%s: %s"):format(util.L.WEBSITE, helpUrl)` instead?

File: LibAddonMenu-2.0/LibAddonMenu-2.0/controls/editbox.lua (hunk 10 above)
Comment (Baertram) [reply LibAddonMenu-2.0/LibAddonMenu-2.0/controls/editbox.lua:25]: Seems it was something from Scootworks addititions... Will strip it

File: LibAddonMenu-2.0/LibAddonMenu-2.0/LibAddonMenu-2.0.lua (hunk 11 above)
Comment (Baertram) [reply LibAddonMenu-2.0/LibAddonMenu-2.0/LibAddonMenu-2.0.lua:113]: I thought having 1 variable would be less often call the string.format stuff. Then I realized that the localization is done further down in the files and not at the beginning, so the access to util.L and L wa snot given at the top.
So whatever you want... :-)

File: LibAddonMenu-2.0/LibAddonMenu-2.0/controls/texture.lua [hunk 12]
```diff
@@ -41,5 +42,7 @@ function LAMCreateControl.texture(parent, textureData, controlName)
         texture:SetHandler("OnMouseExit", ZO_Options_OnMouseExit)
     end
 
+    LAM.util.CreateFAQTexture(control, descriptionData, controlName, LEFT, nil, RIGHT, 0, 0)
```
Comment (sirinsidiator) [inline LibAddonMenu-2.0/LibAddonMenu-2.0/controls/texture.lua:45]: this should receive textureData instead of descriptionData.

File: LibAddonMenu-2.0/LibAddonMenu-2.0/controls/button.lua [hunk 13]
```diff
@@ -85,6 +87,8 @@ function LAMCreateControl.button(parent, buttonData, controlName)
         control:UpdateDisabled()
     end
 
+    LAM.util.CreateFAQTexture(button, buttonData, controlName, LEFT, nil, RIGHT, 0, 0)
```
[the remark this answers is not in the document]
Comment (Baertram) [reply LibAddonMenu-2.0/LibAddonMenu-2.0/controls/button.lua:90]: FAQControl is not created in CreateBaseControl but only from CreateLabelAndContainerControl

[the remark this answers is not in the document]
Comment (sirinsidiator) [reply LibAddonMenu-2.0/LibAddonMenu-2.0/controls/button.lua:90]: Ah sorry. Didn't notice the hidden lines in the diff view.
